File: algorithms/algorithms-cplusplus/leetcode/Question_0005_Longest_Palindromic_Substring.cc
#include <string>

using namespace std;

class Solution {
 public:
  string longestPalindrome(string s) {
    int start = 0, len = 0;
    for (int i = 0; i < s.size(); i++) {
      expandAroundCenter(s, i, i, start, len);
      expandAroundCenter(s, i, i + 1, start, len);
    }
    return s.substr(start, len);
  }

  void expandAroundCenter(string& s, int l, int r, int& start, int& len) {
    while (l >= 0 && r < s.size() && s[l] == s[r]) {
      l--; r++;
    }
    if (r - l - 1 > len) {
      len = r - l - 1;
      start = l + 1;
    }
  }
};

File: algorithms/algorithms-python/leetcode/Question_0005_Longest_Palindromic_Substring.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        res = [0, 0]
        for i in range(len(s)):
            self.help(s, i, i, res)
            self.help(s, i, i + 1, res)
        return s[res[0]:res[0] + res[1]]


    def help(self, s, l, r, res):
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        if r - l - 1 > res[1]:
            res[1] = r - l - 1
            res[0] = l + 1
